### PR TITLE
vim-patch:8.2.4644: redrawing too often when 'relativenumber' is set 

### LINKED ISSUE #9984: Relative line numbering causes lag and slow down in neovim
- `nvim --version`:
> NVIM v0.3.1
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-P1bWuS/neovim-0.3.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=2 -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim-P1bWuS/neovim-0.3.1/build/config -I/build/neovim-P1bWuS/neovim-0.3.1/src -I/usr/include -I/build/neovim-P1bWuS/neovim-0.3.1/build/src/nvim/auto -I/build/neovim-P1bWuS/neovim-0.3.1/build/include
Compiled by team+vim@tracker.debian.org
> Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"
>   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
> Run :checkhealth for more info

- Operating system/version: Ubuntu 18.10 64-bit with GNOME 3.30.1
- Terminal name/version: Konsole version 18.04.3
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:edit (some large file minimum 150-200 lines)
:set rnu

```

### Actual behaviour
Relative line numbering causes lag and slow down although in the editor, this is very noticeable in Konsole but even in the standard Terminal there is a slow down although not as substantial

### Expected behaviour
The editor to continue running smoothely with no perceptable change in scrolling, editing, etc.

### PR DESCRIPTION
**Pull request by lewis6991**
Fixes #9984

To test use `set nu rnu wd=10 rdb=nodelta`.

### CODE REVIEW

File: src/nvim/buffer_defs.h [hunk 1]
```diff
@@ -1212,6 +1212,8 @@ struct window_S {
   colnr_T w_old_visual_col;         ///< last known start of visual part
   colnr_T w_old_curswant;           ///< last known value of Curswant
 
+  linenr_T w_old_cursor_lnum_rnu;   ///< used to track when cursor moves horizontally
```
Comment (zeertzjq) [inline src/nvim/buffer_defs.h:1215]: vertically?

Comment (lewis6991) [reply src/nvim/buffer_defs.h]: 🧠

File: src/nvim/ops.c [hunk 2]
```diff
@@ -2835,8 +2835,8 @@ static void op_yank_reg(oparg_T *oap, bool message, yankreg_T *reg, bool append)
     curr->y_size = j;
     xfree(reg->y_array);
   }
-  if (curwin->w_p_rnu) {
-    redraw_later(curwin, SOME_VALID);  // cursor moved to start
+  if (curwin->w_p_rnu && yanklines > 1) {
+    redraw_later(curwin, VALID);  // cursor moved to start
```
Comment (zeertzjq) [inline src/nvim/ops.c:2839]: This breaks 'cursorline' redrawing only when 'relativenumber' is set. This is strange.

Comment (lewis6991) [reply src/nvim/ops.c]: Fixed

File: src/nvim/ops.c [hunk 3]
```diff
@@ -2835,8 +2835,11 @@ static void op_yank_reg(oparg_T *oap, bool message, yankreg_T *reg, bool append)
     curr->y_size = j;
     xfree(reg->y_array);
   }
-  if (curwin->w_p_rnu) {
-    redraw_later(curwin, SOME_VALID);  // cursor moved to start
+  if (curwin->w_p_rnu && yanklines > 1) {
+    // TODO(lewis6991): for some unknown reason if 'cursoreline' is set,
```
Comment (zeertzjq) [inline src/nvim/ops.c:2839]:
```suggestion
    // TODO(lewis6991): for some unknown reason if 'cursorline' is set,
```

**Comment by zeertzjq** [comment]
I'd prefer being careful with this PR.

**Comment by lewis6991** [comment]
Careful in what sense? You mean not targeting this for 0.7?

**Comment by zeertzjq** [comment]
I just want to try to figure out more about the cursorline redrawing problem.

**Comment by lewis6991** [comment]
Sometimes the best solution is to let master gang find the issues for us 😉 , but that probably means excluding this from 0.7.

File: src/nvim/ops.c [hunk 4]
```diff
@@ -2835,8 +2835,11 @@ static void op_yank_reg(oparg_T *oap, bool message, yankreg_T *reg, bool append)
     curr->y_size = j;
     xfree(reg->y_array);
   }
-  if (curwin->w_p_rnu) {
-    redraw_later(curwin, SOME_VALID);  // cursor moved to start
+  if (curwin->w_p_rnu && yanklines > 1) {
+    // TODO(lewis6991): for some unknown reason if 'cursorline' is set,
+    // redrawing with VALID here breaks and we must use SOME_VALID.
+    // See test 'always updated. vim-patch:8.1.0849' in test/functional/ui/highlight_spec.lua
+    redraw_later(curwin, curwin->w_p_cul ? SOME_VALID : VALID);  // cursor moved to start
```
Comment (zeertzjq) [inline src/nvim/ops.c:2842]: I've figured out something about this redrawing problem. Using `VALID` here causes `update_screen()` to be called before `validate_cursor()`, and `update_screen()` adds `VALID_WCOL` to `wp->w_valid`, which prevents `curs_rows()` from being called later.

Comment (zeertzjq) [reply src/nvim/ops.c:2842]: Perhaps this call to `redraw_later()` isn't necessary at all?

Comment (zeertzjq) [reply src/nvim/ops.c:2842]: Oh yes, this call is unnecessary after patch 8.1.0374

Comment (zeertzjq) [reply src/nvim/ops.c:2842]: <https://github.com/vim/vim/pull/9971>

File: src/nvim/screen.c [hunk 5]
```diff
@@ -1601,6 +1601,9 @@ static void win_update(win_T *wp, DecorProviders *providers)
    * End of loop over all window lines.
    */
 
+  if (wp->w_p_rnu && buf == curwin->w_buffer) {
+    wp->w_old_cursor_lnum_rnu = curwin->w_cursor.lnum;
```
Comment (zeertzjq) [inline src/nvim/screen.c:1605]: This causes `'relativenumber'` not to be redrawn with splits and `nvim_win_set_cursor()`:
1. Open two splits with the same buffer
2. Move the cursor to a different line in one window
3. Use `nvim_win_set_cursor()` to move the cursor in the other window to a different line
4. Use `nvim_win_set_cursor()` to move the cursor in the other window to the same line

**Comment by zeertzjq** [comment]
I've finally figured out the real cause of the cursorline redrawing: it was a bug in `win_update()`, and was fixed in Vim in <https://github.com/vim/vim/commit/e7a74d53754765f22ef8ce71c915bb669d5f7f3f>.

File: src/nvim/screen.c (hunk 5 above)
Comment (zeertzjq) [reply src/nvim/screen.c:1605]: Using `curwin` here doesn't seem to make sense.

File: src/nvim/change.c [hunk 6]
```diff
@@ -287,7 +287,7 @@ static void changed_common(linenr_T lnum, colnr_T col, linenr_T lnume, long xtra
       }
 
       // Relative numbering may require updating more.
-      if (wp->w_p_rnu) {
+      if (wp->w_p_rnu && xtra != 0) {
         redraw_later(wp, SOME_VALID);
```
Comment (lewis6991) [inline src/nvim/change.c:291]: I'm not convinced we need this at all.

Comment (lewis6991) [reply src/nvim/change.c]: Ah we do.

File: src/nvim/buffer_defs.h [hunk 7]
```diff
@@ -1211,6 +1211,8 @@ struct window_S {
   colnr_T w_old_visual_col;         ///< last known start of visual part
   colnr_T w_old_curswant;           ///< last known value of Curswant
 
+  linenr_T w_last_cursor_lnum_rnu;  ///< used to track when cursor moves vertically
```
Comment (zeertzjq) [inline src/nvim/buffer_defs.h:1214]: Moving this between `w_last_cursorline` and `w_last_cursormoved` looks better, also the comment can be clearer:
```suggestion
  linenr_T w_last_cursor_lnum_rnu;  ///< cursor lnum when 'rnu' was last redrawn
```

**Review by zeertzjq — APPROVED** [review]
The code change look fine and it significantly reduces redrawing when editing text in Insert mode. I just think it is very easy to apply this to Vim first, then mark the patch as ported.

**Comment by lewis6991** [comment]
Raised https://github.com/vim/vim/pull/10040

**Comment by zeertzjq** [comment]
Hmm, Vim doesn't have `rdb`.

**Comment by lewis6991** [comment]
Hmm, try setting 'wd' to a negative number?

**Comment by zeertzjq** [comment]
I mean your top comment in that PR.

**Comment by lewis6991** [comment]
Yes, I know. If you look at #9984, `set wd=-N` became`set wd=N rdb=nodelta`

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**